### ISSUE TITLE
Should use `mmap` for model loading

### ISSUE THREAD

**Issue by l29ah** [ggerganov/llama.cpp#91]
So it doesn't create an extra copy in RAM and lives in the kernel page cache happily, loading instantly on subsequent runs.

**Comment by apaz-cli** [comment]
@ggerganov I'm working on putting together the PR. Almost done.

I don't know anything about the order that ggml accesses the weights in. Would you say that it's sequential? If so, there's also `madvise()`.

**Comment by apage43** [comment]
you probably *don't* want to use madvise+`MADV_SEQUENTIAL`, as in addition to increasing the amount of readahead it also causes pages to be evicted after they've been read - the entire model is going to be executed at least once per output token and read all the weights, `MADV_SEQUENTIAL` would potentially kick them all out and reread them repeatedly.

what may be more appropriate is to use `MADV_WILLNEED` on the whole model to kick off paging it all in without needing to wait for it to finish - but `mmap` can be tricky and you would probably want to make it an option rather than the default as it may not be a perf improvement on all setups and can wind up being *slower* than regular I/O due to causing lots of TLB shootdowns - you would want to benchmark it, as its not unlikely you may be trading improved time-to-first-token for worse overall throughput

**Comment by ggerganov** [comment]
@jart 
Thanks for stepping in. I will share briefly an idea that might be useful. Just mind I haven't looked into details of the discussion - will do in a few days once things cool off a bit here.

I think `ggml_context` is extremely well fit for `mmap` if I understand how it works. The `ggml_context` uses an externally provided buffer of memory with a pre-determined size:

https://github.com/ggerganov/llama.cpp/blob/721311070e31464ac12bef9a4444093eb3eaebf7/main.cpp#L569-L575

All tensors and model parameters are "emplaced" into this buffer. There are no extra allocations ocuring.
Once you load the model, you can simply dump the memory buffer provided to `ggml_context` and next time, you can simply load this buffer instead of constructing it. Everything should work.

**Comment by nazthelizard122** [comment]
> All tensors and model parameters are "emplaced" into this buffer. There are no extra allocations ocuring.

In other words, there is a set memory cap? 
Like many other people I have basically no experience in AIs or memory buffers
I'm just a guy pushing buttons until either something explodes or an ai becomes self aware lol

**Comment by bitRAKE** [comment]
For loading from a physical drive resharding the larger models to a single file might help, imho. Whereas loading multiple files in parallel would be slower.

https://github.com/jankais3r/LLaMA_MPS/blob/main/reshard.py

**Comment by apaz-cli** [comment]
I'm getting ready to take another swing at it. My idea of what to do so far:

1. Create functions in `utils.h` called `llama_load_buffer()`, `llama_save_buffer()`, and `llama_destroy_buffer()`. These will `mmap()` (or just `malloc()` and read), save, and `munmap()` (or just `free()`) the buffers respectively. So, files saved on one machine can't necessarily be loaded by another. These files would be stored in some folder, and have the names of the original files. Either in `models`, or `/tmp`, or a new folder. This will also hopefully be useful for implementing saving the model state.

2. Add a new command line argument that tells `llama_model_load()` to look in this cache folder first. If it finds the file, `llama_load_buffer()` the file to get your `ggml_init_params`. Then do whatever else needs to be done (initialize the vocab, get hparams, etc) and exit the function. If the argument is present, call `llama_save_buffer()` first. Also, call `llama_destroy_buffer()` at the appropriate location.

I can do 1, I'll submit a PR for that shortly, but it isn't super clear to me how memory is laid out so that I can do 2. In particular, I'm wondering about the "whatever else needs to be done" part. I'm certain that I'm missing something, and it probably wouldn't be obvious to me what I'm breaking even after hours of monkeying.

**Comment by jart** [comment]
My concern with doing that is, wouldn't it effectively double the disk usage? LLaMA is big enough that folks are already likely to be stretched thin on disk space after creating the second copy needed to quantize the model. I'm still working on studying the codebase to determine what exactly are the transformation that need to be made at runtime. For example, if it's just a bunch of float16's on disk, and we're using a bunch of float16's in memory, then I don't see why the buffer field of these tensors couldn't just be populated with pointers to the appropriate positions in the file. Unless of course it needed to be reshaped or shifted to meet things like AVX alignment requirements. In that case, we'd ideally want to modify the quantizer script so that it generates something suitable for our purposes, so that only a single conversion step needs to be performed.

**Comment by bitRAKE** [comment]
This is the way I was thinking about it:

After the model loads and the prompt is tokenized, create a hash of the context. If that hash exists in the cache directory (and flag is set), load state.

Saving the state has two modes, imho: post-prompt and at end of generation. The post-prompt mode allows jump-starting the model. End saving allows to start-up where one left off.

As to the memory organization - I'd leave that to existing code. The hash would act as pseudo-verification of that it's okay to load a buffer of bytes. The model and prompt would need to be the same, maybe even other options.

Just throwing out some head-space. Haven't starting coding anything.

**Comment by jart** [comment]
Wait. What problem are we trying to solve here exactly? Are we trying to (1) eliminate the three second startup delay? Or are we trying to (2) store the changes made to memory back to disk? Because if your goal is to solve (2) then the only thing you need to save are the random seed and the prompt, since that would restore the state deterministically. Right now I'm focusing on (1) since having fast mmap() loading would not change llama.cpp's behavior, and would instead simply make it go faster. If you want (2) then this change could be the stepping stone you need. All you'd have to do is change `MAP_PRIVATE` to be `MAP_SHARED` instead, and whatever mutations are made to the tensors in memory will be transparently remembered on disk. However that's orthogonal to my intended goals at the moment.

**Comment by bitRAKE** [comment]
I did not intent to expand the meaning of the thread. (2) should probably be addressed elsewhere.

**Comment by apaz-cli** [comment]
@jart It would double the disk usage, yes. But so does converting the model, and so does quantizing it. I think people are prepared for this.

You're right though in that the scripts that convert the model are probably the best way to do this. I was only thinking about implementing a cache for `ggml_init_params` as originally suggested. Ideally though, everything should just be one call, for everything from vocab/hparams to model weights.

**Comment by jart** [comment]
So I added some logging statements to track the read() operations that are happening. It's 200k+ lines that look like this:

```
moving 0x640 bytes from offset 0x4a607 to offset 0 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4ac47 to offset 0xc80 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4b287 to offset 0x1900 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4b8c7 to offset 0x2580 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4bf07 to offset 0x3200 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4c547 to offset 0x3e80 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4cb87 to offset 0x4b00 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4d1c7 to offset 0x5780 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4d807 to offset 0x6400 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4de47 to offset 0x7080 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4e487 to offset 0x7d00 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4eac7 to offset 0x8980 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4f107 to offset 0x9600 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4f747 to offset 0xa280 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x4fd87 to offset 0xaf00 (n_dims=2 n_parts=2)
moving 0x640 bytes from offset 0x503c7 to offset 0xbb80 (n_dims=2 n_parts=2)
```

All it's doing is (1) reshaping and (2) aligning the data in the file. That's why llama.cpp takes several seconds to start. It wouldn't make sense to cache a bunch of memcpy() operations. The quickest thing we could do is introduce a third conversion step that creates a new file format, where the data is in the appropriate shape and alignment ahead of time. Then we could work our way backwards through the conversion tools, to reduce the number of pipeline chores from 3 to 1.

**Comment by jart** [comment]
Here's another reason why this issue is so important. I just ran the 13B model with F16C on my workstation with 32GB of RAM. The model, once loaded, comes very close to hitting the physical memory limit, using maybe ~30GB peak RSS. Bringing memory up to the edge of swapping effectively compounds tragedy, since the kernel reacts by dropping its file caches. If we were using mmap() then the kernel would know that the loaded pages and the file pages are the same thing. But since we're copying the memory, the file cache goes away, and loading ends up taking a minute long each time.

**Comment by jart** [comment]
@apaz-cli Have you attempted implementing yet the thing you proposed? It might work if you use `MAP_FIXED` when reloading it, since GGML appears to allocate objects with pointers too.

**Comment by apaz-cli** [comment]
@jart I have no idea how to support that in a portable way. I haven't dug too deep into it. I'm halfway through implementing part 1.

The troubling thing is actually the default implementation for opening files with the C/C++ stdlib. There is no portable way in C++11 to check the size of a file or binary stream, not even with `fseek()`/`seekg()` and `ftell()`/`tellg()`. C++17 resolves this with `std::filesystem`, but other versions of the standard are out of luck. You have to guess, and resize/copy if you're wrong. Which seems not acceptable. The other way to do it is to read all the bytes of the file once just to get the size, and then do it again. This seems also not acceptable. Unless the compiler is somehow magically able to see through it. I haven't checked, but it doesn't seem that likely.

See [this link to the C standard](https://port70.net/~nsz/c/c11/n1570.html#7.21.9.2p3). The C++ standard says the same about it's own streams.

Although it's UB, the `fseek()`/`ftell()` dance is a classic, and is supported on *almost* all platforms. So we could just do it anyway.

**Comment by apage43** [comment]
the `mmap` operation itself is going to have its own portability issues, supporting all platforms on a first pass with no `#ifdef`s is unlikely here - `mmap` also requires the fd being mapped to actually be a file on a filesystem, which is *probably* implied if its seekable, but `fstat` (or equivalent) is probably the better way to check for that and get the size at the same time

**Comment by jart** [comment]
I've implemented a working prototype for UNIX systems. https://github.com/ggerganov/llama.cpp/commit/5b8023d935401072b73b63ea995aaae040d57b87 

Your LLaMA AGI models will now load instantly without any user visible latency.


https://user-images.githubusercontent.com/49262/225831096-fa53d227-84a4-4dd2-8e62-4630c16a481e.mp4

The above video should be all the proof you need. 

I did this by writing an append-only malloc() that lets us transactionally capture heap allocations, so they can be restored on subsequent runs.  It's a ~200 LOC change that only took me a few hours and it worked the first time. Much easier than the alternative, which likely would have entailed serializing C++ STL objects. This change could be productionized as it stands. I'd need to add the WIN32 mmap() code. I'd also need to store the flags and possibly file timestamps for comparison in the serialized object too, since right now the state change can only happen when `magic.dat` is deleted. We'd also want to put this behind a flag.

However, I still firmly believe this change is not the right thing to do. The file format should be fixed so that it's aligned and doesn't need to be reshaped. Doing that will take 1k+ lines of refactorings, plus a data migration, plus changes to the GGML API. I don't think it's possible to use the ggml_init_params::mem_buffer field, because that memory region contains pointers. That makes it basically the same as my malloc() capturing code, except less generalized. If you wanted to mmap() that field in a portable way, you'd have to do what linkers do, and apply fixups to all the pointers. What I thought might make more sense, is doing a `tensor->data = mmap()` from a given file offset for each tensor (since I'm assuming there aren't that many of them?)

I'll also note that the gains here are mostly due to not copying memory anymore, and better cooperation with the kernel's page manager. We unfortunately aren't getting any additional gains from lazy page loading, since this is a dense model. To generate a single token, every single page in the model file needs to be loaded. What this means is that first runs that load from spinning disk are still going to be slow, even though the average case has greatly improved. I don't view that as a problem, since having the better cooperation with the page manager ensures that the kernel file caches are much less likely to be evicted.

**Comment by ggerganov** [comment]
@jart 

>  I don't think it's possible to use the ggml_init_params::mem_buffer field, because that memory region contains pointers.

Sorry, I missed that.

> What I thought might make more sense, is doing a tensor->data = mmap() from a given file offset for each tensor (since I'm assuming there aren't that many of them?)

This is possible for non-sharded models. For example, if you take a look at the [gpt-2](https://github.com/ggerganov/ggml/blob/4c2f924553312c490e79e6e1739c6f4aa9bbd450/examples/gpt-2/) example, the model loading is a straight read from the file into each tensor:

https://github.com/ggerganov/ggml/blob/4c2f924553312c490e79e6e1739c6f4aa9bbd450/examples/gpt-2/main.cpp#L329-L331

However, the larger LLaMA models ( >7B ) are split into parts, and one has to merge them. Each tensor is split across all the parts. Either by rows or by columns. So the reading logic becomes complicated because of that:

https://github.com/ggerganov/llama.cpp/blob/367946c668757532deed929e1d78673c6ac6bcb8/main.cpp#L457-L503

We could create a combined `ggml` model file as part of the setup process, but this way we go back to the issue of needing double disk space.

Anyway, impressive stuff!

**Comment by IkerAriz** [comment]
@ggerganov

> However, the larger LLaMA models ( >7B ) are split into parts, and one has to merge them. Each tensor is split across all the parts. Either by rows or by columns. So the reading logic becomes complicated because of that:

Perhaps the merging can be addressed with multiple mmaps stitched together into a contiguous region as described here:

https://stackoverflow.com/a/34560306

**Comment by jart** [comment]
@apaz-cli @apage43 Earlier you both raised concerns about portability. POSIX code is standardized and portable. I see however that our CI system requires that we support MSVC. I don't have access to MSVC right now. But I've tried my best to implement a set of polyfills for both WIN32 and POSIX.1 in the following change: 0b5448a3a485eb962916ce9579e335084a940f9c Could you please take a look?

**Comment by moonshadow565** [comment]
I did a change like this as well by first swapping std::ifstream to custom file reader that reads out of mmap.
Which btw as consequence showed that one of the .read calls actually fails when reading the data!
It only really works for 7B model because file format is non-optimal for mmap.

+1 on changing file format to merge all data and allow for mmap

**Comment by apaz-cli** [comment]
@jart I'm more lamenting at the absurdity that there's no portable (C++11) way to find the size of a file. It truly baffles me. On posix there's `fstat`. On Windows there's `GetFileSize` / `GetFileSizeEx`. But there are other platforms that are not windows or posix, and they would no longer be supported.

Sadly, I do not have access to a windows machine or MSVC either.

**Comment by jart** [comment]
What are those platforms? I would have assumed C++11 would be less portable than POSIX.

**Comment by heshpdx** [comment]
This should apply to whisper.cpp as well, right? Since it is for a ggml model...

**Comment by tarruda** [comment]
Another approach to make it run instantly without reloading the weights every time is by using a client/server model, which if what I've done in #278. This is how it works:

The model is loaded during server startup, then each connection is served in a forked child, and using a simple protocol you can pass some of the command-line parameters to customize that session and interact with the model over the socket.

It is as if you are instantly spawning llama.cpp with preloaded model, only instead of running the program from the scratch, you open a TCP connection to interact with the program.

**Comment by jart** [comment]
@tarruda What value does your TCP server offer, other than solving a problem that's already being solved by mmap()?

**Comment by tarruda** [comment]
> @tarruda What value does your TCP server offer, other than solving a problem that's already being solved by mmap()?

Replied [here](https://github.com/ggerganov/llama.cpp/pull/278#issuecomment-1475914534).

**Comment by ggerganov** [comment]
@tarruda I think we should not conflate the 2 features: mmap-based model loading and TCP server mode. The first one is more of a "technical superiority" feature, which I think is very important to prioritise and add as a default feature in the codebase. The latter is more of an example, enabling larger accessibility and extending the audience of the project. In my view, it has lower priority at this point, since we are currently trying to lay a good technical foundation - make sure the computations are correct, optimize the performance, fix quantization, etc. The examples will be much easier to build on top of that a bit later.

The way I see it is for the TCP server mode, we first need to finish the C-style API (#77, which I hope to do in the following days) and after that, the proposed TCP changes will become a `llama.cpp` example for building a TCP server (similar to the `whisper.cpp` examples). There are many other ways to make a server, but this will be one of them and it will be available out-of-the box from this project. So in that sense, it is good that we already have a prototype thanks to your changes.

@jart regarding the Win32 compatibility of `mmap()` - if it turns out to be too big of hassle to support it, I think it is fully OK to have instant model loading just for Unix OSes and on Windows we can simply fallback to normal loading.

**Comment by nicknitewolf** [comment]
Just a note, I have compiled the mmap branch on Windows using MSVC and it loads my 7B model just fine. Do you have anything specific you want me to test @jart ?

**Comment by nicknitewolf** [comment]
> > Just a note, I have compiled the mmap branch on Windows using MSVC and it loads my 7B model just fine. Do you have anything specific you want me to test @jart ?
> 
> Could you please describe step by step how you compiled the `mmap` branch? Because right now I am struggling to make this crappy winapi work as intended. I have found several critical bugs in file mapping code but after fixing them it doesn't seem to be loading\saving the model in `magic.dat` correctly. First run is okay, but once the magic is loaded, the pointers for the model and the vocab are simply not readable... I suspect this is because model and vocab are created using `new` operator which generates pointers outside of the memory mapping address range. I tried to allocate memory for them using custom malloc, but it doesn't work, the first access to the vocab crushes everything.

Build Env:
- Windows 11 22621.1413

Tools:
- cmake 3.26.0
- MSBuild version 17.5.0+6f08c67f3 for .NET Framework

Commands:
```
cmake -S . -B build/
cmake --build build/ --config Release -j8
```

**Comment by oKatanaaa** [comment]
@nicknitewolf I followed the steps and it seems it doesn't work afterall. I mean it does run, but on subsequent runs the model is still being loaded which should not happen (loading part must be skipped).

Guess I'll make a PR with my fixes so that other people can take a look as well.

@jart There is something wrong with accessing addresses that are generated in lines 980-981. They can't be read on subsequent runs. Not sure how this works on Linux, but on Win those pointers just contain garbage on sebsequent runs. Again, tried to use custom malloc - no luck. Turned out there was no mmap call for such small allocations so I tried to force it by removing if statement in 185 (not sure what this `if` does, but anyway) - still no luck.

**Comment by nicknitewolf** [comment]
> @nicknitewolf I followed the steps and it seems it doesn't work afterall. I mean it does run, but on subsequent runs the model is still being loaded which should not happen (loading part must be skipped).
> 
> Guess I'll make a PR with my fixes so that other people can take a look as well.
> 
> @jart There is something wrong with accessing addresses that are generated in lines 980-981. They can't be read on subsequent runs. Not sure how this works on Linux, but on Win those pointers just contain garbage on sebsequent runs. Again, tried to use custom malloc - no luck. Turned out there was no mmap call for such small allocations so I tried to force it by removing if statement in 185 (not sure what this `if` does, but anyway) - still no luck.

Do you have any other build tools on your system, e.g. (mingw, msys, etc)? Those build tools in PATH might conflict with MSVC

**Comment by oKatanaaa** [comment]
@nicknitewolf only MSVC.

**Comment by tmm1** [comment]
This may be a useful reference in getting mmap compat working correctly on Windows: https://github.com/edsrzf/mmap-go

**Comment by nicknitewolf** [comment]
[vsconfig.txt](https://github.com/ggerganov/llama.cpp/files/11021384/vsconfig.txt)
My .vsconfig file if you are interested

**Comment by oKatanaaa** [comment]
@nicknitewolf embarrassing to admit, but I forgot to switch the branch to `mmap`. The unfortunate thing though `mmap` branch does not compile at all. Please make sure you switched the branch.

Here is the full log of compilation steps:
[log.txt](https://github.com/ggerganov/llama.cpp/files/11023160/log.txt)

Those are the errors I have already faced earlier. Still need to work out the problem with serializing pointer.

**Comment by oKatanaaa** [comment]
Made a PR (#341) with some adjustments for the Win implementation of mmap just to make it compile at least. So anyone using Visual Studio or having experience in mmap for Win could take a look.

**Comment by slaren** [comment]
Not sure if anybody has tried this approach yet, but I found that single-part models (ie. 7B) can be easily mmap'ed by skipping the reads of the tensors and replacing them with a pointer to the mmap'ed region. Doing this my load times drop to ~15ms.

This is never going to work for multi-part models since some of the tensors are split in different files, but then again I don't think that any mmap approach is going to be able to work with that either. If we exported multi-part tensors into a single file with contiguous tensors this could be fixed. A potential issue may be that tensors may end at an unaligned memory address, but this could also be fixed with some tweaks to the conversion script.

Proof of concept here: https://github.com/slaren/llama.cpp/commit/fc685122f95f212d11c7ce07cf5201c8ac752925

**Comment by CoderRC** [comment]
There is a way to check if mmap exists using _POSIX_MAPPED_FILES in the POSIX standard, it says that from mmap man page, and I have created my own library that implements mmap using mingw32 that makes this project maintainable for windows. It is possible to compile the program using library from https://github.com/CoderRC/libmingw32_extended, make changes like in https://github.com/ggerganov/llama.cpp/pull/564 and the specific make command below:
make LDFLAGS='-D_POSIX_MAPPED_FILES -lmingw32_extended'

**Comment by jart** [comment]
I just reviewed #341 which I think will help bring us closer to ensuring MSVC users will gain the benefits of mmap() too. 

> Not sure if anybody has tried this approach yet, but I found that single-part models (ie. 7B) can be easily mmap'ed by skipping the reads of the tensors and replacing them with a pointer to the mmap'ed region. Doing this my load times drop to ~15ms.

Wow @slaren. That's what I was hoping for originally. I determined earlier it wasn't possible, for the reasons described in this comment: https://github.com/ggerganov/llama.cpp/issues/91#issuecomment-1472893422 I might have been wrong since I can't recall if I gleaned that information (about things like AVX alignment) from the 13B model though. I'm going to try your proof of concept and see what happens. Even if it only works for 7B, your suggestion is something we could do *by default* for 7B model users and it'd make things much better for them.

---

**Side Note:** I have time to focus solely on solving this issue over the next few days, and that's what I intend to do. In addition to being here on GitHub, anyone is welcome to join me on the "Redbean" Discord server https://discord.gg/AqSvHf4u (channel "ai" where I'm also jart) if you share my personal fondness for low latency collaboration. In the future though, @ggerganov will decide the most appropriate chat venue for those of us helping with his projects.

**Comment by jart** [comment]
I've just tested @slaren's change locally. It works great for 7B (which only has 1-dimensional tensors), loads instantly, and has zero performance regression on evaluation. My only worry is that `tensor->data` is no longer aligned to `GGML_MEM_ALIGN` because the file offsets aren't aligned, Does anyone know if that matters? It runs fine for me, unaligned, on AVX, SSE, and Apple M1 is fine too. I can't speak for other ISAs like POWER though.

**Comment by l29ah** [comment]
> Proof of concept here: [slaren@fc68512](https://github.com/slaren/llama.cpp/commit/fc685122f95f212d11c7ce07cf5201c8ac752925)

Tried it out on the single-part `alpaca-13B-ggml/ggml-model-q4_0.bin` i've grabbed from a torrent, and it works like a charm, thank you!

**Comment by slaren** [comment]
I can try to integrate the Windows patch on my branch and cleanup the code a bit so that at least it can be used with 7B for now.

@ggerganov what do you think would be the best way to modify ggml to allow it to create tensors without reserving memory? (like the `ggml_nomem()` hack in https://github.com/slaren/llama.cpp/commit/c13eaf38ba7bc992e91bd54b121f514d9147e1f2)

**Comment by ggerganov** [comment]
@jart 
> I've just tested @slaren's change locally. It works great for 7B (which only has 1-dimensional tensors), loads instantly, and has zero performance regression on evaluation. My only worry is that `tensor->data` is no longer aligned to `GGML_MEM_ALIGN` because the file offsets aren't aligned, Does anyone know if that matters? It runs fine for me, unaligned, on AVX, SSE, and Apple M1 is fine too. I can't speak for other ISAs like POWER though.

Last time I researched this topic, I concluded that there is no real difference between aligned and non-aligned memory memory. Probably this mattered in the past, but nowadays I think the differences are negligible.

I think `ggml.c` uses only non-aligned SIMD intrinsics, unless I missed some recent change on this. So if that is all true, I think it should not really matter if the memory buffer is aligned to `GGML_MEM_ALIGN` or not.

And even if I am wrong, given that all tensors in the model have sizes that are multiples of 32 I think it wouldn't be difficult to guarantee the alignment, right?

@slaren 
> I can try to integrate the Windows patch on my branch and cleanup the code a bit so that at least it can be used with 7B for now.
> 
> @ggerganov what do you think would be the best way to modify ggml to allow it to create tensors without reserving memory? (like the `ggml_nomem()` hack in [slaren@c13eaf3](https://github.com/slaren/llama.cpp/commit/c13eaf38ba7bc992e91bd54b121f514d9147e1f2))

It has to be a new variable in the `struct ggml_init_params`:

https://github.com/ggerganov/llama.cpp/blob/2a98bc18ea34dbf15f261a0df37080e588a189d1/ggml.h#L314-L320

Name it `bool no_alloc` or whatever you prefer

**Comment by jart** [comment]
> And even if I am wrong, given that all tensors in the model have sizes that are multiples of 32 I think it wouldn't be difficult to guarantee the alignment, right?

I can't imagine it'd be difficult at all. I think the main thing that misaligns the format, is those nul-terminated token strings at the beginning of the file. If you do a roundup(offset, 32) after producing those, then you might luck out and everything else will magically align.

**Comment by ggerganov** [comment]
Totally - at this point we can easily make changes to the `ggml` model format and not worry too much.
Having a 32-byte rounded header sounds like a good idea anyway - regardless if we really need it or not

**Comment by Ar57m** [comment]
@ggerganov is possible to you guys make mmap work on gptj(pyg6b)?

**Comment by apaz-cli** [comment]
@Ar57m Why would it not be working for a specific model? Do you have an error message to share? We need more to go on. Also, this is not the right place. You should look up the error message in the issues, and if you can't find it, create a new one.

Your other interaction on Github asks if it's possible to support loading big models on low memory devices. Yes and no. Memory mapping requires that you have enough memory to map into. So if you're running out of RAM, consider allocating swap space. This allows the OS to use the reserved disk space as though it were RAM. It isn't RAM, so expect it to run much slower. But it could make the difference between the program running and not running.

**Comment by l29ah** [comment]
> Memory mapping requires that you have enough memory to map into.

It doesn't. It's just like swap.

**Comment by apaz-cli** [comment]
Enough virtual memory. Also there has to be enough physical memory backing it, or it will fault. So you are technically correct. Possibly the best kind of correct, but this is pedantics.

**Comment by Ar57m** [comment]
@apaz-cli I'm trying to run pyg6b q4_0 on termux(Android), when the model is loading it crashes termux(because it's using too much ram), idk I think it's because it's not mmaping, I heard somewhere that gptj doesn't support mmap. I can run almost any size Llama based models(very slowly big models) with mmap on.
